### ISSUE TITLE
[Feature Request] WAN Failover Status In InfluxDB

### ISSUE THREAD

**Issue by ircmaxell** [unpoller/unpoller#424]
I have recently setup multiple WAN support on my UDM-Pro. Currently, the only data that I have in InfluxDB through Unpoller is the basic infor for each wan port (name, IP, MAC, up status, enabled status, bytes, speed, rx/tx info), but nothing about the active uplink.

I can view fallback status via unifi_alarm table, but that leaves some to be desired:

```
Gateway[<--snip-->:a6] WAN iface eth9 transition to state failover
Gateway[<--snip-->:a6] WAN iface eth8 transition to state active
Gateway[<--snip-->:a6] WAN iface eth9 transition to state active
Gateway[<--snip-->:a6] WAN iface eth8 transition to state inactive
```

To enable this, I'd suggest one of three (or perhaps 2/all 3) options:

1. Adding an "is_uplink" flag to the usg_wan_ports table to allow filtering / displaying that info
2. Adding a new table usg_uplinks which has the information for the active uplink (including IP address and name of the active uplink, uplink source, etc)
3. Adding a field to usg to identify which wan port is active: uplink, uplink_wan, etc. 

Running `unifi-poller -j devices` shows that the data is present.

 ```
     "wan1": {
        "uplink_ifname": "eth8",
        "max_speed": 1000,
        "type": "ethernet",
        "autoneg": true,
        "enable": true,
        "flowctrl_rx": false,
        "flowctrl_tx": false,
        "full_duplex": true,
        "is_uplink": true,
        "mac": "<--snip-->:ae",
        "mac_table": [
          {
            "age": 116361,
            "authorized": true,
            "hostname": "<--snip-->",
            "ip": "<--snip-->.1",
            "lastReachable": 20,
            "mac": "<--snip-->:d8"
          }
        ],
        "media": "GE",
        "name": "eth8",
        "num_port": 1,
        "port_idx": 9,
        "port_poe": false,
        "rx_broadcast": 0,
        "rx_bytes": 401960855639,
        "rx_dropped": 0,
        "rx_errors": 0,
        "rx_multicast": 0,
        "rx_packets": 330381935,
        "rx_rate": 17638422,
        "rx_rate-max": 20141410,
        "speed": 1000,
        "speed_caps": 1048623,
        "tx_broadcast": 0,
        "tx_bytes": 83155981573,
        "tx_dropped": 0,
        "tx_errors": 0,
        "tx_multicast": 0,
        "tx_packets": 139164025,
        "tx_rate": 86231,
        "tx_rate-max": 1124109,
        "up": true,
        "ifname": "eth8",
        "tx_bytes-r": 10778,
        "rx_bytes-r": 2204802,
        "bytes-r": 2215580,
        "ip": "<--snip-->.45",
        "netmask": "255.255.255.0",
        "dns": [
          "10.1.0.10",
          "10.1.0.11"
        ]
      },
      "wan2": {
        "uplink_ifname": "eth9",
        "max_speed": 1000,
        "type": "ethernet",
        "autoneg": true,
        "enable": true,
        "flowctrl_rx": false,
        "flowctrl_tx": false,
        "full_duplex": true,
        "is_uplink": false,
        "mac": "<--snip-->:af",
        "mac_table": [
          {
            "age": 512280,
            "authorized": true,
            "hostname": "?",
            "ip": "<--snip-->.1",
            "lastReachable": 68,
            "mac": "<--snip-->:7c"
          }
        ],
       "media": "SFP+",
        "name": "eth9",
        "num_port": 1,
        "port_idx": 10,
        "port_poe": false,
        "rx_broadcast": 0,
        "rx_bytes": 34913005,
        "rx_dropped": 0,
        "rx_errors": 0,
        "rx_multicast": 0,
        "rx_packets": 621580,
        "rx_rate": 366,
        "rx_rate-max": 2953,
        "sfp_found": true,
        "sfp_part": "<--snip-->",
        "sfp_serial": "<--snip-->",
        "sfp_vendor": "TP-Link",
        "speed": 1000,
        "speed_caps": 1048864,
        "tx_broadcast": 0,
        "tx_bytes": 32169199,
        "tx_dropped": 0,
        "tx_errors": 0,
        "tx_multicast": 0,
        "tx_packets": 599783,
        "tx_rate": 366,
        "tx_rate-max": 2878,
        "up": true,
        "ifname": "eth9",
        "tx_bytes-r": 45,
        "rx_bytes-r": 45,
        "bytes-r": 90,
        "ip": "<--snip-->.243",
        "netmask": "255.255.255.0",
        "dns": [
          "10.1.0.10",
          "10.1.0.11"
        ]
      },
      "uplink": {
        "drops": 3,
        "ip": "<--snip-->.45",
        "latency": 2,
        "name": "eth8",
        "nameservers": [
          "10.1.0.10",
          "10.1.0.11"
        ],
        "netmask": "255.255.255.0",
        "num_port": 1,
        "rx_bytes": 401995768644,
        "rx_dropped": 0,
        "rx_errors": 0,
        "rx_multicast": 0,
        "rx_packets": 331003515,
        "speedtest_lastrun": 1661848193,
        "speedtest_ping": 3,
        "speedtest_status": "Success",
        "tx_bytes": 83188150772,
        "tx_dropped": 0,
        "tx_errors": 0,
        "tx_packets": 139763808,
        "up": true,
        "uptime": 500470,
        "xput_down": 906.813,
        "xput_up": 923.246,
        "port_idx": 9,
        "media": "GE",
        "speed": 1000,
        "full_duplex": true,
        "rx_rate": 17638422,
        "tx_rate": 86231,
        "max_speed": 1000,
        "type": "wire",
        "tx_bytes-r": 10824,
        "rx_bytes-r": 2204848,
        "uplink_source": "legacy",
        "bytes-r": 2215672
      },
      "last_wan_ip": "<--snip-->.45",
```

**Comment by brngates98** [comment]
I have some ideas for this -- working on implementing this for prometheus, influx, datadog... found some really good data here...

**Comment by brngates98** [comment]
I've created a PR to add comprehensive WAN metrics support:

**Library PR:** unpoller/unifi#189 - Adds WAN data structures and API methods

This will enable monitoring of:
- **WAN Failover Status** - Which interface is ACTIVE vs BACKUP
- **Load Balancing** - Weights, priorities, mode (weighted/failover-only)
- **Uptime Statistics** - Uptime percentage per WAN interface
- **Peak Usage** - Peak download/upload percentages
- **Service Provider Info** - ISP name, city, ASN

Once the library PR is merged, I have a follow-up PR ready for unpoller that will export these metrics to:
- **Prometheus** - `unpoller_wan_*` metrics
- **InfluxDB** - `wan` measurement with all fields and tags
- **Datadog** - `unpoller.wan.*` metrics

This should fully address the feature request for WAN Failover Status tracking!

**Comment by brngates98** [comment]
**Update:** The unpoller PR is now open as well!

- **Library PR:** unpoller/unifi#189 - WAN data structures and API methods
- **Unpoller PR:** #925 - WAN metrics export to Prometheus, InfluxDB, and Datadog

Once both PRs are merged, WAN failover status and all related metrics will be available in all three output backends.

**Comment by dsymonds** [comment]
Hey, I've updated to the v2.27.0 release that picked those changes up, but I'm not seeing any WAN metrics on the /metrics output. Was their any config on the unpoller side to enable this? Or anything to do on the Unifi controller side (a UDM-SE for me)?

**Comment by brngates98** [comment]
> Hey, I've updated to the v2.27.0 release that picked those changes up, but I'm not seeing any WAN metrics on the /metrics output. Was their any config on the unpoller side to enable this? Or anything to do on the Unifi controller side (a UDM-SE for me)?

investigating this now

**Comment by dsymonds** [comment]
I turned on debugging for unpoller, and get this:

```
unpoller  | 2026/01/30 00:44:54.286436 updateweb.go:226: [DEBUG] Fetching WAN enriched configuration for site default
unpoller  | 2026/01/30 00:44:54.286531 updateweb.go:226: [DEBUG] unifi.GetWANEnrichedConfiguration(https://10.0.0.1): creating request: parse "https://10.0.0.1/proxy/network/v2/api/site/%s/wan/enriched-configuration": invalid URL escape "%s/" (continuing)
```
(10.0.0.1 is my UDM-SE).

**Comment by dsymonds** [comment]
To try to help rule out any issue on my end, I'm running `UniFi Network Application 10.0.162`, and all the other unpoller metrics seem fine.

**Comment by brngates98** [comment]
issue was on my end -- have a incoming fix im dumb sometimes

**Comment by dsymonds** [comment]
Thanks for working on this in the first place, and for a quick fix!

**Comment by brngates98** [comment]
> Thanks for working on this in the first place, and for a quick fix!

<img width="2018" height="396" alt="Image" src="https://github.com/user-attachments/assets/25421ee3-0898-4936-8f8c-1002370d6c65" />its working locally

**Comment by brngates98** [comment]
testing few other things and ill PR

**Comment by brngates98** [comment]
## Fix Available

I've submitted a fix for this issue in the unifi library: https://github.com/unpoller/unifi/pull/190

### Problem

The WAN API methods were passing the raw API path constant (containing `%s` placeholder) directly to `GetData`, which caused an `invalid URL escape "%s/"` error when using the Remote API mode via `api.ui.com`. This prevented WAN metrics from being collected.

### Solution

- Use `fmt.Sprintf` to properly format site name into API paths before making requests
- Switch from `GetData` to `GetJSON` for endpoints that return top-level arrays
- Add defensive checks to detect unformatted `%s` in paths early
- Applied fix consistently across all WAN API methods

### Tested & Working

After the fix, all WAN metrics are now being collected successfully:

- `unpoller_wan_failover_priority` - Priority (1=primary, 2=backup)
- `unpoller_wan_load_balance_weight` - Load balancing weight
- `unpoller_wan_provider_download_kbps` / `upload_kbps` - ISP speeds
- `unpoller_wan_uptime_percentage` - Uptime percentage
- `unpoller_wan_peak_download_percent` / `upload_percent` - Peak utilization
- `unpoller_wan_service_provider_asn` - ISP ASN with `isp_name`, `isp_city` labels
- And more (smartq, magic, vlan, max rates, etc.)

The WAN failover status is now available through these metrics, including the `is_uplink` concept via the failover priority and load balance weight metrics.

**Comment by brngates98** [comment]
the error was applicable to local and remote connections I just failed to mention that but I did test

**Comment by brngates98** [comment]
Now to come up with some dashboard ideas

**Comment by twiebe** [comment]
> ### Tested & Working
> After the fix, all WAN metrics are now being collected successfully:
> 
> * `unpoller_wan_failover_priority` - Priority (1=primary, 2=backup)
> * `unpoller_wan_load_balance_weight` - Load balancing weight
> * `unpoller_wan_provider_download_kbps` / `upload_kbps` - ISP speeds
> * `unpoller_wan_uptime_percentage` - Uptime percentage
> * `unpoller_wan_peak_download_percent` / `upload_percent` - Peak utilization
> * `unpoller_wan_service_provider_asn` - ISP ASN with `isp_name`, `isp_city` labels
> * And more (smartq, magic, vlan, max rates, etc.)
> 
> The WAN failover status is now available through these metrics, including the `is_uplink` concept via the failover priority and load balance weight metrics.

@brngates98 , could you elaborate how exactly you determine the failover **status** from these metrics? i'm looking at them via the prometheus endpoint, but when my failover connection (LTE) kicks in, `unpoller_wan_failover_priority` does not change and stays like this:
```
# HELP unpoller_wan_failover_priority WAN failover priority (lower is higher priority)
# TYPE unpoller_wan_failover_priority gauge
unpoller_wan_failover_priority{site_name="",source="",wan_id="6943dad48a89532cda42e666",wan_load_balance_type="weighted",wan_name="Kabel",wan_networkgroup="WAN",wan_type="static"} 1
unpoller_wan_failover_priority{site_name="",source="",wan_id="6943dad48a89532cda42e667",wan_load_balance_type="failover-only",wan_name="LTE",wan_networkgroup="WAN2",wan_type="static"} 2
```

while in the webinterface i see

<img width="607" height="144" alt="Image" src="https://github.com/user-attachments/assets/6ac96acf-fe6d-4e0a-9ec4-0237bff677c3" />

this is with unpoller 2.33.0.
